GKE Node Maintenance Doc (#2)
appuio/rancher-docs: docs/modules/ROOT/pages/how-tos/gke_node_maintenance.adoc

= GKE Node Maintenance

== Prerequisites

* Access to the customer `gke-terraform` repository

> == Preparation

Reviewer: New line missing after title

* Login to the GKE console for the cluster you want to maintain / upgrade (in case you need to manually intervene)
* Check out the customer's `gke-terraform` repository

== Node maintenance

> GKE allows you to auto-upgrade your nodes. See link:https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-upgrades[Cluster upgrades] and link:https://cloud.google.com/kubernetes-engine/docs/how-to/node-auto-upgrades[Auto-upgrading nodes] for more details.

Reviewer: ```suggestion
GKE allows you to auto-upgrade your nodes. See https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-upgrades[Cluster upgrades] and https://cloud.google.com/kubernetes-engine/docs/how-to/node-auto-upgrades[Auto-upgrading nodes] for more details.
```

Auto-Upgrade is enabled for the nodes by default.

. Change Kubernetes Version
+
No release channel is selected and therefore the version is set static. Change `cluster_version` in your `gke-terraform` repository.

. Verify Auto-Upgrade Settings
+
The node auto-upgrade is controlled by the following variables:
+
* `maintenance_start_time`: Time window specified for recurring maintenance operations in RFC3339 format (default: `05:00`)
* `maintenance_end_time`: Time window specified for recurring maintenance operations in RFC3339 format (default: ``)
* `maintenance_recurrence`: Time window specified for recurring maintenance operations in RFC3339 format (default: ``)
* `worker_max_surge`: The number of additional worker nodes that can be added to the node pool during an upgrade
* `worker_max_unavailable`: The number of worker nodes that can be simultaneously unavailable during an upgrade
* `builder_max_surge` The number of additional builder nodes that can be added to the node pool during an upgrade
* `builder_max_unavailable`: The number of builder nodes that can be simultaneously unavailable during an upgrade
+

> This results in a daily 4h (default minimum value) maintenance windows.

Reviewer: ```suggestion
This results in a daily 4h (default minimum value) maintenance window.
```

> = GKE Node Maintenance

Reviewer 2: For someone not familiar with GKE, this is more confusing than helpful. At least I do not get it.

I understand, that raising the Kubernetes Version is a intentional change done within the Terraform repository of that cluster. Will this have an imediate effect? The Auto-Upgrade suggests no. But then I am supposed to log into the Google Console so I can act when things go thouth. This suggests that I have to babysit the process.

So my assumption is this:

I change the Kubernetes Version in code. Then I ensure the terraform apply got executed (pipeline). And herer my job is done.

The actuall upgrade will happen automatically within the next maintenance window.

If things go south, monitoring should alert. Here comes the need for logging into the Google Console. So mentioning this would go more into a troubleshoot section which could be a how to on its own.

Author: thx for the feedback. I'll try to rewrite it and make it more clear.

Reviewer 2: Yes thanks. This now makes sense to me.

== Prerequisites

* Access to the customer `gke-terraform` repository

== Preparation

* Login to the GKE console for the cluster you want to maintain / upgrade (in case you need to manually intervene)
* Check out the customer's `gke-terraform` repository

== Node maintenance

GKE allows you to auto-upgrade your nodes. See https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-upgrades[Cluster upgrades] and https://cloud.google.com/kubernetes-engine/docs/how-to/node-auto-upgrades[Auto-upgrading nodes] for more details.
Auto-Upgrade is enabled for the nodes by default.

. Change Kubernetes Version
+

> No release channel is selected and therefore the version is set static. Change `cluster_version` in your `gke-terraform` repository.

Reviewer 3: ```suggestion
No release channel is selected and therefore the version is set statically. Change the parameter `cluster_version` in your `gke-terraform` repository to change the Kubernetes version.
```

Author: :+1:

. Verify Auto-Upgrade Settings
+
The node auto-upgrade is controlled by the following variables:
+

> * `maintenance_start_time`: Time window specified for recurring maintenance operations in RFC3339 format (default: `05:00`)

Reviewer 3: ```suggestion
* `maintenance_start_time`: Start time for the recurring maintenance operations in https://www.ietf.org/rfc/rfc5545.txt[RFC-5545] format (default: `05:00`)
```

According to the GKE docs, this is not an RFC3339 time, but a RFC5545 `DTSTART`. (https://cloud.google.com/kubernetes-engine/docs/how-to/maintenance-windows-and-exclusions).

Author: for the gcloud cli, yes, but api spec says RFC3339 See https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/projects.locations.clusters#dailymaintenancewindow & https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/projects.locations.clusters#recurringtimewindow & https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/projects.locations.clusters#timewindow

I'll clarify and update the documentation to make it more clear

> * `maintenance_end_time`: Time window specified for recurring maintenance operations in RFC3339 format (default: ``)

Reviewer 3: Copy-pasted parameter description?

Author: this is copy-pasted from terraform module documentation: https://registry.terraform.io/modules/terraform-google-modules/kubernetes-engine/google/latest/submodules/beta-private-cluster
but yes I made some copy-paste errors

> * `maintenance_recurrence`: Time window specified for recurring maintenance operations in RFC3339 format (default: ``)

Reviewer 3: This description also looks copy-pasted and doesn't match the parameter name. A corrected description should mention that the format is a RFC5545 `RRULE`, as documented in https://cloud.google.com/kubernetes-engine/docs/how-to/maintenance-windows-and-exclusions.

Author: See comment above, I'll try to clarify this better

* `worker_max_surge`: The number of additional worker nodes that can be added to the node pool during an upgrade
* `worker_max_unavailable`: The number of worker nodes that can be simultaneously unavailable during an upgrade
* `builder_max_surge` The number of additional builder nodes that can be added to the node pool during an upgrade
* `builder_max_unavailable`: The number of builder nodes that can be simultaneously unavailable during an upgrade
+

> This results in a daily 4h (default minimum value) maintenance window.

Reviewer 3: I don't quite understand how leaving `maintenance_recurrence` empty results in a daily maintenance window. Linking to the relevant GKE documentation (https://cloud.google.com/kubernetes-engine/docs/how-to/maintenance-windows-and-exclusions) might help here.

It would be nice to have an example for a weekly maintenance window, so it's more readily understandable how values for the mentioned parameters look like.

Author: I'll clarify and make an example with the next commit

= GKE Node Maintenance

== Node maintenance

GKE allows you to auto-upgrade your Kubernetes control-plane and nodes.
See https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-upgrades[Cluster upgrades] and https://cloud.google.com/kubernetes-engine/docs/how-to/node-auto-upgrades[Auto-upgrading nodes] for more details.
Auto-Upgrade is enabled for the nodes by default.
GKE allows you to specify maintenance windows and exclusions to control when an auto-upgrade is executed.
See https://cloud.google.com/kubernetes-engine/docs/how-to/maintenance-windows-and-exclusions[Configuring maintenance windows and exclusions]

> === Change Kubernetes Version

Reviewer: I somehow don't get what this section wants to tell the reader in regards to the node maintenance. Either it needs clarification or should be deleted.

Author: basically it should be a tl;dr for the google docs. I don't want to copy past content.
I can try to clarify it a bit more (without just copy/paste content from the google docs) 
or just put the important links and then let the user read it by himself?

Author: I have made some sub header below "Node maintenance" with first for a better structure:
* a tl;dr & links to the important docs from google for the node maintenance 
* and then the actual doing to initiate the version upgrade form the cluster and the nodes

I hope this clairfy's it a bit more?

= GKE Node Maintenance

== Prerequisites

* Access to the customer `gke-terraform` repository

== Preparation

* Check out the customer's `gke-terraform` repository

== Node maintenance

> GKE allows you to auto-upgrade your kubernetes control plane and nodes. See https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-upgrades[Cluster upgrades] and https://cloud.google.com/kubernetes-engine/docs/how-to/node-auto-upgrades[Auto-upgrading nodes] for more details.

Reviewer 2: ```suggestion
GKE allows you to auto-upgrade your kubernetes control plane and nodes.
See https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-upgrades[Cluster upgrades] and https://cloud.google.com/kubernetes-engine/docs/how-to/node-auto-upgrades[Auto-upgrading nodes] for more details.
```

Auto-Upgrade is enabled for the nodes by default.

> GKE allows you to specify maintenance windows and exclustions to control when an auto-upgrade is executed. See https://cloud.google.com/kubernetes-engine/docs/how-to/maintenance-windows-and-exclusions[Configuring maintenance windows and exclusions]

Reviewer 2: ```suggestion
GKE allows you to specify maintenance windows and exclustions to control when an auto-upgrade is executed.
See https://cloud.google.com/kubernetes-engine/docs/how-to/maintenance-windows-and-exclusions[Configuring maintenance windows and exclusions]
```

. Change Kubernetes Version
+

> No release channel is selected and therefore the version is set statically. Change the parameter `cluster_version` in your `gke-terraform` repository to change the Kubernetes version.

Reviewer 2: ```suggestion
No release channel is selected and therefore the version is set statically.
Change the parameter `cluster_version` in your `gke-terraform` repository to change the Kubernetes version.
```

> No release channel is selected and therefore the version is set statically. Change the parameter `cluster_version` in your `gke-terraform` repository to change the Kubernetes version.

Reviewer 2: ```suggestion
No release channel is selected and therefore the version is set statically.
Change the parameter `cluster_version` in your `gke-terraform` repository to change the Kubernetes version.
```

+
* `maintenance_start_time`: Time window specified for daily or recurring maintenance operations in RFC3339 format (default: `05:00`)
* `maintenance_end_time`: Time window specified for recurring maintenance operations in RFC3339 format (default: ``)
* `maintenance_recurrence`: Frequency of the recurring maintenance window in RFC5545 format. (default: ``)
* `worker_max_surge`: The number of additional worker nodes that can be added to the node pool during an upgrade
* `worker_max_unavailable`: The number of worker nodes that can be simultaneously unavailable during an upgrade
* `builder_max_surge` The number of additional builder nodes that can be added to the node pool during an upgrade
* `builder_max_unavailable`: The number of builder nodes that can be simultaneously unavailable during an upgrade
+
Google GKE allows for a https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/projects.locations.clusters#Cluster.DailyMaintenanceWindow[simple daily] or a https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/projects.locations.clusters#recurringtimewindow[recurring maintenance] window.
+

> When `maintenance_recurrence` is not set, this automaticly results in a daily maintenance window. See terraform https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/v12.3.0/modules/beta-private-cluster/main.tf#L175[module's sourcecode] for details on this distinction.

Reviewer 2: ```suggestion
When `maintenance_recurrence` is not set, this automaticly results in a daily maintenance window.
See terraform https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/v12.3.0/modules/beta-private-cluster/main.tf#L175[module's sourcecode] for details on this distinction.
```

Only `maintenance_start_time` is required for the daily maintenance window to use.
A minimum duration of 4h is requrired, see https://cloud.google.com/kubernetes-engine/docs/how-to/maintenance-windows-and-exclusions[Configuring maintenance windows and exclusions]
The duration in the case of the daily mainenance window cannot be set with Terraform and therefore defaults to 4h.
+
By setting `maintenance_start_time`, `maintenance_end_time` and `maintenance_recurrence` you define a recurring maintenance window.
Example for a recurring maintenance window:
+

> ----

Reviewer 2: ```suggestion
[source,yaml]
----
```

This should give some syntax highlighting in the output.
